### PR TITLE
Convert drill code to TypeScript

### LINKED ISSUE #26949: Migrate drill code to TypeScript

### PR DESCRIPTION
**Pull request by kulyk**
Closes #26949, epic #26836

Refreshes and consolidates viz click actions and drills types and converts all the existing drill code to TypeScript.
That's the first step to update our click actions popover UI/UX. No changes in functionality expected

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/27165)

### CODE REVIEW

File: frontend/src/metabase/modes/types.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  url: () => string;
+};
+
+export type RegularClickActionEffect =
+  | ReduxClickAction
+  | QuestionChangeClickAction
+  | PopoverClickAction
+  | UrlClickAction;
+
+type RegularClickAction = ClickActionBase & RegularClickActionEffect;
```
Comment (ranquild) [inline frontend/src/metabase/modes/types.ts:86]: Could we remove this type if we extend specific action types?

**Review by ranquild — APPROVED** [review]

File: frontend/src/metabase/modes/types.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  section: string;
+  icon?: string;
+  buttonType: ClickActionButtonType;
+  default?: boolean;
+  defaultAlways?: boolean;
+  tooltip?: string;
+  extra?: () => Record<string, unknown>;
+};
+
+type ReduxClickAction = {
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/modes/types.ts]: Done 👍

File: frontend/src/metabase/modes/types.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+export type ClickActionButtonType =
+  | "formatting"
+  | "horizontal"
+  | "info"
+  | "token"
+  | "token-filter"
+  | "sort";
+
+export type ClickActionBase = {
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/modes/types.ts:50]: Yeah, that'd be the best option, but I'm worried the `metabase-lib` occurrences might make it harder. Overall, I believe `metabase-lib` should be unaware of `ClickObject`. Let me try to explore this in a different branch

File: frontend/src/metabase/modes/types.ts (hunk 1 above)
Comment (kulyk) [reply frontend/src/metabase/modes/types.ts]: We're using it down below for the [`AlwaysDefaultClickAction` type](https://github.com/metabase/metabase/blob/a61ed7d2ae4053e5e00ac1015dded683b4f3cf6a/frontend/src/metabase/modes/types.ts#L85).

Any action with `defaultAlways: true` is treated as one. Having an action like that, Metabase will ignore all the other possible drills and immediately execute the always default one. AFAIK it was introduced for the click behavior drill, which should act exactly that way. TBH it feels this behavior "overwrite" should happen on the dashboard level, but fixing this feels out of the scope of this change.

Any click action (URL / redux action / etc.) can be an "always default". Important point here is that for the "always default" action, we can skip the UI-related properties like section, icon, title, etc. as there's no popup to show them in. So we extend the `RegularClickAction` type [here](https://github.com/metabase/metabase/blob/a61ed7d2ae4053e5e00ac1015dded683b4f3cf6a/frontend/src/metabase/modes/types.ts#L86) and omit all the irrelevant properties. And then the final [`ClickAction` type](https://github.com/metabase/metabase/blob/a61ed7d2ae4053e5e00ac1015dded683b4f3cf6a/frontend/src/metabase/modes/types.ts#L92) is either a regular action that can be presented in a popup OR an "always default" action

I've tried to simplify this a bit and address other feedback, would appreciate if you can take another look at this @ranquild

**Pull request merged**